---
title: "Cov2Comparator"
author: "Kuai Yu"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Cov2Comparator}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction
Cov2Comparator is a package that analyzes SARS covid-2 genome across different 
geographic regions. It allows easy retrieval of SARS COV-2 genome 
. It can perform multiple sequence alignments and 
phylogenetic tree. Plot of multiple sequence alignments and 
phylogenetic tree can be generated for interpretation. This document gives a 
tour of Cov2Comparator package.

see `help(package = "Cov2Comparator")` for more information and 
references provided by `citation("Cov2Comparator")`

To download **Cov2Comparator**, use the following commands:

```r
require("devtools")
devtools::install_github("quick2063706271/Cov2Comparator", build_vignettes = TRUE)
library("Cov2Comparator")
```

To list all functions available in this package:

```r
ls("package:Cov2Comparator")
```

To list dataset in this package:

``` r
data(package = "Cov2Comparator")
```

## Components
There are 7 functions in this package

### Loading and retrieving sequence from NCBI 
There are three ways to load and retrieve sequence. 
__*getSequenceByRegion*__ allows user to retrieve a SARS-COV2 genome by 
inputting a region which user desired. Current supports input from 
(Canada, Italy, Wuhan, USA, Kenya, Bahrain, Germany, Pakistan, Britain, 
Thailand)

Here is an example.
```{r}
# Please make sure you are connected to the internet
wuhanSeq <- getSequenceByRegion("Wuhan")
print(wuhanSeq)
```

__*getSequencesByRegions*__ allows user to retrieve a SARS-COV2 genome by 
inputting multiple regions which user desired. Current supports input from 
(Canada, Italy, Wuhan, USA, Kenya, Bahrain, Germany, Pakistan, Britain, 
Thailand)

```{r}
# Please make sure you are connected to the internet
regions <- c("Wuhan", "Italy", "USA")
seqs <- getSequencesByRegions(regions)
print(seqs)
```

The last way to add is using __*readGenome*__. This function allows user to input fasta file as input, user can also include a nameToRegion file, which contains one to one correspondence of accessionID and region.
```{r}
fastaPath1 <- system.file("extdata", "MN985325.1.fasta", package="Cov2Comparator")
nameToRegionsFile <- system.file("extdata", "nameToCountry.txt", package="Cov2Comparator")
fasta1 <- readGenome(fastaPath1, nameToRegionsFile)
fasta1
```

User do not to provide nameToRegion file. In that case, users are recommanded to
modify the name by themselves

```{r}
fastaPath1 <- system.file("extdata", "MN985325.1.fasta", package="Cov2Comparator")
nameToRegionsFile <- system.file("extdata", "nameToCountry.txt", package="Cov2Comparator")
fasta1 <- readGenome(fastaPath1)
print(fasta1)
# After modifying name
cat("\nAfter modifying name\n\n")
names(fasta1) <- c("MN985325.1 USA")
print(fasta1)
```

After loading the data, we can perform a multiple sequence alignment using 
__*multiplSeqAlign*__. This functions is derived from __*msa::msa*__ 
```{R}
seqs
align <- multiplSeqAlign(seqs, algorithm = 'ClustalW')
align
```
We can plot the alignment using __*plotAlignment*__

```{r}
p <- plotAlignment(align, 'align.fasta', refSequence = 'NC_045512.2 Wuhan')
p
```

After that, multiple sequence alignment can be used to generate phylogenetic 
tree using __*createTree*__
```{r}
tree <- createTree(align)
```

Tree can be ploted using __*plotTree*__

```{r}
plotTree(tree,
         name = "demo Alignment",
         showRegionName = TRUE)
```



